src/brogue/IO.c

```diff
@@ -3168,19 +3339,44 @@ void messageWithoutCaps(char *msg, enum messageFlags flags) {
         }
     }*/
 
-    for (i = MESSAGE_LINES - 1; i >= 1; i--) {
-        messageConfirmed[i] = messageConfirmed[i-1];
-        strcpy(displayedMessage[i], displayedMessage[i-1]);
+    // Add the message to the archive, bumping counts for recent duplicates
+    newMessage = true;
+
+    // For each at most MESSAGE_ARCHIVE_ENTRIES - 1 past entries..
+    for (i = 1; i < MESSAGE_ARCHIVE_ENTRIES; i++) {
```

Comment:
I'm struggling to understanding what this loop body does, or the function in general

Reply:
This is a pre-existing function, whose purpose before and after is to insert new messages into the message archive.

On [`master`][0], we can inspect its anatomy and conclude:
* It manages the "confirmed" (is bolded or not) state of the normal three-line message display, and shifts up to make room for the new message, which it then inserts.
* It inserts the message into the archive, which is a circular buffer.
* Get acknowledgement (show `--MORE--` and wait) from the player if required

[0]: https://github.com/tmewett/BrogueCE/blob/master/src/brogue/IO.c#L3154-L3192

On this branch:
* It first walks the archive, looking for a suitable message to collapse the new one into.  If one is found, its count is bumped.  This is the loop you have called out.
* If no suitable message to collapse into was found, instead initialize and insert the incoming message into the archive (again, dealing with the circular buffer) as a new one.  Only at this point, when we know a new row will be shown in the three-line display and in the archive, do we shift up.
* Again get acknowledgement if required, though `requireAcknowledgment` has become `REQUIRE_ACKNOWLEDGMENT`.

`updateMessageDisplay` is also replaced by `displayRecentMessages(false)`.  The `false` in particular means that newly arrived messages will be unconfirmed (shown in bold.)  It occurs to me that we can probably drop the `messageConfirmed` shifting now.